// @vendure/common/lib/shared-constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CREATING_VENDURE_APP = exports.DEFAULT_AUTH_TOKEN_HEADER_KEY = exports.ROOT_COLLECTION_NAME = exports.CUSTOMER_ROLE_DESCRIPTION = exports.CUSTOMER_ROLE_CODE = exports.SUPER_ADMIN_USER_PASSWORD = exports.SUPER_ADMIN_USER_IDENTIFIER = exports.SUPER_ADMIN_ROLE_DESCRIPTION = exports.SUPER_ADMIN_ROLE_CODE = exports.DEFAULT_CHANNEL_CODE = exports.SHOP_API_PATH = exports.ADMIN_API_PATH = exports.API_PORT = void 0;
/*
 * This file contains constants which are shared between more than one sub-module
 * e.g. values required by both the server and admin-ui.
 */
exports.API_PORT = 3000;
exports.ADMIN_API_PATH = 'admin-api';
exports.SHOP_API_PATH = 'shop-api';
exports.DEFAULT_CHANNEL_CODE = '__default_channel__';
exports.SUPER_ADMIN_ROLE_CODE = '__super_admin_role__';
exports.SUPER_ADMIN_ROLE_DESCRIPTION = 'SuperAdmin';
exports.SUPER_ADMIN_USER_IDENTIFIER = 'superadmin';
exports.SUPER_ADMIN_USER_PASSWORD = 'superadmin';
exports.CUSTOMER_ROLE_CODE = '__customer_role__';
exports.CUSTOMER_ROLE_DESCRIPTION = 'Customer';
exports.ROOT_COLLECTION_NAME = '__root_collection__';
exports.DEFAULT_AUTH_TOKEN_HEADER_KEY = 'vendure-auth-token';
exports.CREATING_VENDURE_APP = 'CREATING_VENDURE_APP';
