// @vendure/core/dist/entity/custom-entity-fields.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomZoneFields = exports.CustomUserFields = exports.CustomTaxRateFields = exports.CustomTaxCategoryFields = exports.CustomStockLocationFields = exports.CustomShippingMethodFieldsTranslation = exports.CustomShippingMethodFields = exports.CustomSellerFields = exports.CustomRegionFieldsTranslation = exports.CustomRegionFields = exports.CustomPromotionFieldsTranslation = exports.CustomPromotionFields = exports.CustomProductVariantFieldsTranslation = exports.CustomProductVariantFields = exports.CustomProductOptionGroupFieldsTranslation = exports.CustomProductOptionGroupFields = exports.CustomProductOptionFieldsTranslation = exports.CustomProductOptionFields = exports.CustomProductFieldsTranslation = exports.CustomProductFields = exports.CustomPaymentMethodFieldsTranslation = exports.CustomPaymentMethodFields = exports.CustomOrderLineFields = exports.CustomOrderFields = exports.CustomGlobalSettingsFields = exports.CustomFulfillmentFields = exports.CustomFacetValueFieldsTranslation = exports.CustomFacetValueFields = exports.CustomFacetFieldsTranslation = exports.CustomFacetFields = exports.CustomCustomerGroupFields = exports.CustomCustomerFields = exports.CustomCollectionFieldsTranslation = exports.CustomCollectionFields = exports.CustomChannelFields = exports.CustomAssetFields = exports.CustomAdministratorFields = exports.CustomAddressFields = void 0;
class CustomAddressFields {
}
exports.CustomAddressFields = CustomAddressFields;
class CustomAdministratorFields {
}
exports.CustomAdministratorFields = CustomAdministratorFields;
class CustomAssetFields {
}
exports.CustomAssetFields = CustomAssetFields;
class CustomChannelFields {
}
exports.CustomChannelFields = CustomChannelFields;
class CustomCollectionFields {
}
exports.CustomCollectionFields = CustomCollectionFields;
class CustomCollectionFieldsTranslation {
}
exports.CustomCollectionFieldsTranslation = CustomCollectionFieldsTranslation;
class CustomCustomerFields {
}
exports.CustomCustomerFields = CustomCustomerFields;
class CustomCustomerGroupFields {
}
exports.CustomCustomerGroupFields = CustomCustomerGroupFields;
class CustomFacetFields {
}
exports.CustomFacetFields = CustomFacetFields;
class CustomFacetFieldsTranslation {
}
exports.CustomFacetFieldsTranslation = CustomFacetFieldsTranslation;
class CustomFacetValueFields {
}
exports.CustomFacetValueFields = CustomFacetValueFields;
class CustomFacetValueFieldsTranslation {
}
exports.CustomFacetValueFieldsTranslation = CustomFacetValueFieldsTranslation;
class CustomFulfillmentFields {
}
exports.CustomFulfillmentFields = CustomFulfillmentFields;
class CustomGlobalSettingsFields {
}
exports.CustomGlobalSettingsFields = CustomGlobalSettingsFields;
class CustomOrderFields {
}
exports.CustomOrderFields = CustomOrderFields;
class CustomOrderLineFields {
}
exports.CustomOrderLineFields = CustomOrderLineFields;
class CustomPaymentMethodFields {
}
exports.CustomPaymentMethodFields = CustomPaymentMethodFields;
class CustomPaymentMethodFieldsTranslation {
}
exports.CustomPaymentMethodFieldsTranslation = CustomPaymentMethodFieldsTranslation;
class CustomProductFields {
}
exports.CustomProductFields = CustomProductFields;
class CustomProductFieldsTranslation {
}
exports.CustomProductFieldsTranslation = CustomProductFieldsTranslation;
class CustomProductOptionFields {
}
exports.CustomProductOptionFields = CustomProductOptionFields;
class CustomProductOptionFieldsTranslation {
}
exports.CustomProductOptionFieldsTranslation = CustomProductOptionFieldsTranslation;
class CustomProductOptionGroupFields {
}
exports.CustomProductOptionGroupFields = CustomProductOptionGroupFields;
class CustomProductOptionGroupFieldsTranslation {
}
exports.CustomProductOptionGroupFieldsTranslation = CustomProductOptionGroupFieldsTranslation;
class CustomProductVariantFields {
}
exports.CustomProductVariantFields = CustomProductVariantFields;
class CustomProductVariantFieldsTranslation {
}
exports.CustomProductVariantFieldsTranslation = CustomProductVariantFieldsTranslation;
class CustomPromotionFields {
}
exports.CustomPromotionFields = CustomPromotionFields;
class CustomPromotionFieldsTranslation {
}
exports.CustomPromotionFieldsTranslation = CustomPromotionFieldsTranslation;
class CustomRegionFields {
}
exports.CustomRegionFields = CustomRegionFields;
class CustomRegionFieldsTranslation {
}
exports.CustomRegionFieldsTranslation = CustomRegionFieldsTranslation;
class CustomSellerFields {
}
exports.CustomSellerFields = CustomSellerFields;
class CustomShippingMethodFields {
}
exports.CustomShippingMethodFields = CustomShippingMethodFields;
class CustomShippingMethodFieldsTranslation {
}
exports.CustomShippingMethodFieldsTranslation = CustomShippingMethodFieldsTranslation;
class CustomStockLocationFields {
}
exports.CustomStockLocationFields = CustomStockLocationFields;
class CustomTaxCategoryFields {
}
exports.CustomTaxCategoryFields = CustomTaxCategoryFields;
class CustomTaxRateFields {
}
exports.CustomTaxRateFields = CustomTaxRateFields;
class CustomUserFields {
}
exports.CustomUserFields = CustomUserFields;
class CustomZoneFields {
}
exports.CustomZoneFields = CustomZoneFields;
