// @vendure/core/dist/api/config/generate-active-order-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateActiveOrderTypes = void 0;
const stitch_1 = require("@graphql-tools/stitch");
const graphql_1 = require("graphql");
const errors_1 = require("../../common/error/errors");
const active_order_strategy_1 = require("../../config/order/active-order-strategy");
/**
 * This function is responsible for constructing the `ActiveOrderInput` GraphQL input type.
 * It does so based on the inputs defined by the configured ActiveOrderStrategy defineInputType
 * methods, dynamically building a mapped input type of the format:
 *
 *```
 * {
 *     [strategy_name]: strategy_input_type
 * }
 * ```
 */
function generateActiveOrderTypes(schema, activeOrderStrategies) {
    const fields = {};
    const strategySchemas = [];
    const strategiesArray = Array.isArray(activeOrderStrategies)
        ? activeOrderStrategies
        : [activeOrderStrategies];
    for (const strategy of strategiesArray) {
        if (typeof strategy.defineInputType === 'function') {
            const inputSchema = (0, graphql_1.buildASTSchema)(strategy.defineInputType());
            const inputType = Object.values(inputSchema.getTypeMap()).find((type) => (0, graphql_1.isInputObjectType)(type));
            if (!inputType) {
                throw new errors_1.InternalServerError(`${strategy.constructor.name}.defineInputType() does not define a GraphQL Input type`);
            }
            fields[strategy.name] = { type: inputType };
            strategySchemas.push(inputSchema);
        }
    }
    if (Object.keys(fields).length === 0) {
        return schema;
    }
    const activeOrderInput = new graphql_1.GraphQLInputObjectType({
        name: 'ActiveOrderInput',
        fields,
    });
    const activeOrderOperations = [
        { name: 'activeOrder', isMutation: false },
        { name: 'eligibleShippingMethods', isMutation: false },
        { name: 'eligiblePaymentMethods', isMutation: false },
        { name: 'nextOrderStates', isMutation: false },
        { name: 'addItemToOrder', isMutation: true },
        { name: 'adjustOrderLine', isMutation: true },
        { name: 'removeOrderLine', isMutation: true },
        { name: 'removeAllOrderLines', isMutation: true },
        { name: 'applyCouponCode', isMutation: true },
        { name: 'removeCouponCode', isMutation: true },
        { name: 'addPaymentToOrder', isMutation: true },
        { name: 'setCustomerForOrder', isMutation: true },
        { name: 'setOrderShippingAddress', isMutation: true },
        { name: 'setOrderBillingAddress', isMutation: true },
        { name: 'setOrderShippingMethod', isMutation: true },
        { name: 'setOrderCustomFields', isMutation: true },
        { name: 'transitionOrderToState', isMutation: true },
    ];
    const queryType = schema.getQueryType();
    const mutationType = schema.getMutationType();
    const strategyNames = strategiesArray.map(s => s.name).join(', ');
    const description = `Inputs for the configured ${strategiesArray.length === 1 ? 'ActiveOrderStrategy' : 'ActiveOrderStrategies'} ${strategyNames}`;
    for (const operation of activeOrderOperations) {
        const field = operation.isMutation
            ? mutationType === null || mutationType === void 0 ? void 0 : mutationType.getFields()[operation.name]
            : queryType === null || queryType === void 0 ? void 0 : queryType.getFields()[operation.name];
        if (!field) {
            throw new errors_1.InternalServerError(`Could not find a GraphQL type definition for the field ${operation.name}`);
        }
        // TODO: Figure out a non-hacky way to do this!
        field.args.push({
            name: active_order_strategy_1.ACTIVE_ORDER_INPUT_FIELD_NAME,
            type: activeOrderInput,
            description,
            defaultValue: null,
            extensions: null,
            astNode: null,
            deprecationReason: null,
        });
    }
    return (0, stitch_1.stitchSchemas)({
        subschemas: [schema, ...strategySchemas],
        types: [activeOrderInput],
        typeMergingOptions: { validationSettings: { validationLevel: stitch_1.ValidationLevel.Off } },
    });
}
exports.generateActiveOrderTypes = generateActiveOrderTypes;
