// @vendure/core/dist/common/error/generated-graphql-admin-errors.js
"use strict";
/* eslint-disable */
/** This file was generated by the graphql-errors-plugin, which is part of the "codegen" npm script. */
Object.defineProperty(exports, "__esModule", { value: true });
exports.adminErrorOperationTypeResolvers = exports.SettlePaymentError = exports.RefundStateTransitionError = exports.RefundPaymentIdMissingError = exports.RefundOrderStateError = exports.QuantityTooGreatError = exports.ProductOptionInUseError = exports.PaymentStateTransitionError = exports.PaymentOrderMismatchError = exports.PaymentMethodMissingError = exports.OrderStateTransitionError = exports.OrderModificationStateError = exports.OrderModificationError = exports.OrderLimitError = exports.NothingToRefundError = exports.NoChangesSpecifiedError = exports.NoActiveOrderError = exports.NegativeQuantityError = exports.NativeAuthStrategyError = exports.MultipleOrderError = exports.MissingConditionsError = exports.MimeTypeError = exports.ManualPaymentStateError = exports.LanguageNotAvailableError = exports.ItemsAlreadyFulfilledError = exports.InvalidFulfillmentHandlerError = exports.InvalidCredentialsError = exports.InsufficientStockOnHandError = exports.InsufficientStockError = exports.IneligibleShippingMethodError = exports.GuestCheckoutError = exports.FulfillmentStateTransitionError = exports.FacetInUseError = exports.EmptyOrderLineSelectionError = exports.EmailAddressConflictError = exports.CreateFulfillmentError = exports.CouponCodeLimitError = exports.CouponCodeInvalidError = exports.CouponCodeExpiredError = exports.ChannelDefaultLanguageError = exports.CancelPaymentError = exports.CancelActiveOrderError = exports.AlreadyRefundedError = exports.ErrorResult = void 0;
class ErrorResult {
}
exports.ErrorResult = ErrorResult;
class AlreadyRefundedError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'AlreadyRefundedError';
        this.errorCode = 'ALREADY_REFUNDED_ERROR';
        this.message = 'ALREADY_REFUNDED_ERROR';
        this.refundId = input.refundId;
    }
}
exports.AlreadyRefundedError = AlreadyRefundedError;
class CancelActiveOrderError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'CancelActiveOrderError';
        this.errorCode = 'CANCEL_ACTIVE_ORDER_ERROR';
        this.message = 'CANCEL_ACTIVE_ORDER_ERROR';
        this.orderState = input.orderState;
    }
}
exports.CancelActiveOrderError = CancelActiveOrderError;
class CancelPaymentError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'CancelPaymentError';
        this.errorCode = 'CANCEL_PAYMENT_ERROR';
        this.message = 'CANCEL_PAYMENT_ERROR';
        this.paymentErrorMessage = input.paymentErrorMessage;
    }
}
exports.CancelPaymentError = CancelPaymentError;
class ChannelDefaultLanguageError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'ChannelDefaultLanguageError';
        this.errorCode = 'CHANNEL_DEFAULT_LANGUAGE_ERROR';
        this.message = 'CHANNEL_DEFAULT_LANGUAGE_ERROR';
        this.channelCode = input.channelCode;
        this.language = input.language;
    }
}
exports.ChannelDefaultLanguageError = ChannelDefaultLanguageError;
class CouponCodeExpiredError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'CouponCodeExpiredError';
        this.errorCode = 'COUPON_CODE_EXPIRED_ERROR';
        this.message = 'COUPON_CODE_EXPIRED_ERROR';
        this.couponCode = input.couponCode;
    }
}
exports.CouponCodeExpiredError = CouponCodeExpiredError;
class CouponCodeInvalidError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'CouponCodeInvalidError';
        this.errorCode = 'COUPON_CODE_INVALID_ERROR';
        this.message = 'COUPON_CODE_INVALID_ERROR';
        this.couponCode = input.couponCode;
    }
}
exports.CouponCodeInvalidError = CouponCodeInvalidError;
class CouponCodeLimitError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'CouponCodeLimitError';
        this.errorCode = 'COUPON_CODE_LIMIT_ERROR';
        this.message = 'COUPON_CODE_LIMIT_ERROR';
        this.couponCode = input.couponCode;
        this.limit = input.limit;
    }
}
exports.CouponCodeLimitError = CouponCodeLimitError;
class CreateFulfillmentError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'CreateFulfillmentError';
        this.errorCode = 'CREATE_FULFILLMENT_ERROR';
        this.message = 'CREATE_FULFILLMENT_ERROR';
        this.fulfillmentHandlerError = input.fulfillmentHandlerError;
    }
}
exports.CreateFulfillmentError = CreateFulfillmentError;
class EmailAddressConflictError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'EmailAddressConflictError';
        this.errorCode = 'EMAIL_ADDRESS_CONFLICT_ERROR';
        this.message = 'EMAIL_ADDRESS_CONFLICT_ERROR';
    }
}
exports.EmailAddressConflictError = EmailAddressConflictError;
class EmptyOrderLineSelectionError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'EmptyOrderLineSelectionError';
        this.errorCode = 'EMPTY_ORDER_LINE_SELECTION_ERROR';
        this.message = 'EMPTY_ORDER_LINE_SELECTION_ERROR';
    }
}
exports.EmptyOrderLineSelectionError = EmptyOrderLineSelectionError;
class FacetInUseError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'FacetInUseError';
        this.errorCode = 'FACET_IN_USE_ERROR';
        this.message = 'FACET_IN_USE_ERROR';
        this.facetCode = input.facetCode;
        this.productCount = input.productCount;
        this.variantCount = input.variantCount;
    }
}
exports.FacetInUseError = FacetInUseError;
class FulfillmentStateTransitionError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'FulfillmentStateTransitionError';
        this.errorCode = 'FULFILLMENT_STATE_TRANSITION_ERROR';
        this.message = 'FULFILLMENT_STATE_TRANSITION_ERROR';
        this.fromState = input.fromState;
        this.toState = input.toState;
        this.transitionError = input.transitionError;
    }
}
exports.FulfillmentStateTransitionError = FulfillmentStateTransitionError;
class GuestCheckoutError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'GuestCheckoutError';
        this.errorCode = 'GUEST_CHECKOUT_ERROR';
        this.message = 'GUEST_CHECKOUT_ERROR';
        this.errorDetail = input.errorDetail;
    }
}
exports.GuestCheckoutError = GuestCheckoutError;
class IneligibleShippingMethodError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'IneligibleShippingMethodError';
        this.errorCode = 'INELIGIBLE_SHIPPING_METHOD_ERROR';
        this.message = 'INELIGIBLE_SHIPPING_METHOD_ERROR';
    }
}
exports.IneligibleShippingMethodError = IneligibleShippingMethodError;
class InsufficientStockError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'InsufficientStockError';
        this.errorCode = 'INSUFFICIENT_STOCK_ERROR';
        this.message = 'INSUFFICIENT_STOCK_ERROR';
        this.order = input.order;
        this.quantityAvailable = input.quantityAvailable;
    }
}
exports.InsufficientStockError = InsufficientStockError;
class InsufficientStockOnHandError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'InsufficientStockOnHandError';
        this.errorCode = 'INSUFFICIENT_STOCK_ON_HAND_ERROR';
        this.message = 'INSUFFICIENT_STOCK_ON_HAND_ERROR';
        this.productVariantId = input.productVariantId;
        this.productVariantName = input.productVariantName;
        this.stockOnHand = input.stockOnHand;
    }
}
exports.InsufficientStockOnHandError = InsufficientStockOnHandError;
class InvalidCredentialsError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'InvalidCredentialsError';
        this.errorCode = 'INVALID_CREDENTIALS_ERROR';
        this.message = 'INVALID_CREDENTIALS_ERROR';
        this.authenticationError = input.authenticationError;
    }
}
exports.InvalidCredentialsError = InvalidCredentialsError;
class InvalidFulfillmentHandlerError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'InvalidFulfillmentHandlerError';
        this.errorCode = 'INVALID_FULFILLMENT_HANDLER_ERROR';
        this.message = 'INVALID_FULFILLMENT_HANDLER_ERROR';
    }
}
exports.InvalidFulfillmentHandlerError = InvalidFulfillmentHandlerError;
class ItemsAlreadyFulfilledError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'ItemsAlreadyFulfilledError';
        this.errorCode = 'ITEMS_ALREADY_FULFILLED_ERROR';
        this.message = 'ITEMS_ALREADY_FULFILLED_ERROR';
    }
}
exports.ItemsAlreadyFulfilledError = ItemsAlreadyFulfilledError;
class LanguageNotAvailableError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'LanguageNotAvailableError';
        this.errorCode = 'LANGUAGE_NOT_AVAILABLE_ERROR';
        this.message = 'LANGUAGE_NOT_AVAILABLE_ERROR';
        this.languageCode = input.languageCode;
    }
}
exports.LanguageNotAvailableError = LanguageNotAvailableError;
class ManualPaymentStateError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'ManualPaymentStateError';
        this.errorCode = 'MANUAL_PAYMENT_STATE_ERROR';
        this.message = 'MANUAL_PAYMENT_STATE_ERROR';
    }
}
exports.ManualPaymentStateError = ManualPaymentStateError;
class MimeTypeError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'MimeTypeError';
        this.errorCode = 'MIME_TYPE_ERROR';
        this.message = 'MIME_TYPE_ERROR';
        this.fileName = input.fileName;
        this.mimeType = input.mimeType;
    }
}
exports.MimeTypeError = MimeTypeError;
class MissingConditionsError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'MissingConditionsError';
        this.errorCode = 'MISSING_CONDITIONS_ERROR';
        this.message = 'MISSING_CONDITIONS_ERROR';
    }
}
exports.MissingConditionsError = MissingConditionsError;
class MultipleOrderError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'MultipleOrderError';
        this.errorCode = 'MULTIPLE_ORDER_ERROR';
        this.message = 'MULTIPLE_ORDER_ERROR';
    }
}
exports.MultipleOrderError = MultipleOrderError;
class NativeAuthStrategyError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'NativeAuthStrategyError';
        this.errorCode = 'NATIVE_AUTH_STRATEGY_ERROR';
        this.message = 'NATIVE_AUTH_STRATEGY_ERROR';
    }
}
exports.NativeAuthStrategyError = NativeAuthStrategyError;
class NegativeQuantityError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'NegativeQuantityError';
        this.errorCode = 'NEGATIVE_QUANTITY_ERROR';
        this.message = 'NEGATIVE_QUANTITY_ERROR';
    }
}
exports.NegativeQuantityError = NegativeQuantityError;
class NoActiveOrderError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'NoActiveOrderError';
        this.errorCode = 'NO_ACTIVE_ORDER_ERROR';
        this.message = 'NO_ACTIVE_ORDER_ERROR';
    }
}
exports.NoActiveOrderError = NoActiveOrderError;
class NoChangesSpecifiedError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'NoChangesSpecifiedError';
        this.errorCode = 'NO_CHANGES_SPECIFIED_ERROR';
        this.message = 'NO_CHANGES_SPECIFIED_ERROR';
    }
}
exports.NoChangesSpecifiedError = NoChangesSpecifiedError;
class NothingToRefundError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'NothingToRefundError';
        this.errorCode = 'NOTHING_TO_REFUND_ERROR';
        this.message = 'NOTHING_TO_REFUND_ERROR';
    }
}
exports.NothingToRefundError = NothingToRefundError;
class OrderLimitError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'OrderLimitError';
        this.errorCode = 'ORDER_LIMIT_ERROR';
        this.message = 'ORDER_LIMIT_ERROR';
        this.maxItems = input.maxItems;
    }
}
exports.OrderLimitError = OrderLimitError;
class OrderModificationError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'OrderModificationError';
        this.errorCode = 'ORDER_MODIFICATION_ERROR';
        this.message = 'ORDER_MODIFICATION_ERROR';
    }
}
exports.OrderModificationError = OrderModificationError;
class OrderModificationStateError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'OrderModificationStateError';
        this.errorCode = 'ORDER_MODIFICATION_STATE_ERROR';
        this.message = 'ORDER_MODIFICATION_STATE_ERROR';
    }
}
exports.OrderModificationStateError = OrderModificationStateError;
class OrderStateTransitionError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'OrderStateTransitionError';
        this.errorCode = 'ORDER_STATE_TRANSITION_ERROR';
        this.message = 'ORDER_STATE_TRANSITION_ERROR';
        this.fromState = input.fromState;
        this.toState = input.toState;
        this.transitionError = input.transitionError;
    }
}
exports.OrderStateTransitionError = OrderStateTransitionError;
class PaymentMethodMissingError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'PaymentMethodMissingError';
        this.errorCode = 'PAYMENT_METHOD_MISSING_ERROR';
        this.message = 'PAYMENT_METHOD_MISSING_ERROR';
    }
}
exports.PaymentMethodMissingError = PaymentMethodMissingError;
class PaymentOrderMismatchError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'PaymentOrderMismatchError';
        this.errorCode = 'PAYMENT_ORDER_MISMATCH_ERROR';
        this.message = 'PAYMENT_ORDER_MISMATCH_ERROR';
    }
}
exports.PaymentOrderMismatchError = PaymentOrderMismatchError;
class PaymentStateTransitionError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'PaymentStateTransitionError';
        this.errorCode = 'PAYMENT_STATE_TRANSITION_ERROR';
        this.message = 'PAYMENT_STATE_TRANSITION_ERROR';
        this.fromState = input.fromState;
        this.toState = input.toState;
        this.transitionError = input.transitionError;
    }
}
exports.PaymentStateTransitionError = PaymentStateTransitionError;
class ProductOptionInUseError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'ProductOptionInUseError';
        this.errorCode = 'PRODUCT_OPTION_IN_USE_ERROR';
        this.message = 'PRODUCT_OPTION_IN_USE_ERROR';
        this.optionGroupCode = input.optionGroupCode;
        this.productVariantCount = input.productVariantCount;
    }
}
exports.ProductOptionInUseError = ProductOptionInUseError;
class QuantityTooGreatError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'QuantityTooGreatError';
        this.errorCode = 'QUANTITY_TOO_GREAT_ERROR';
        this.message = 'QUANTITY_TOO_GREAT_ERROR';
    }
}
exports.QuantityTooGreatError = QuantityTooGreatError;
class RefundOrderStateError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'RefundOrderStateError';
        this.errorCode = 'REFUND_ORDER_STATE_ERROR';
        this.message = 'REFUND_ORDER_STATE_ERROR';
        this.orderState = input.orderState;
    }
}
exports.RefundOrderStateError = RefundOrderStateError;
class RefundPaymentIdMissingError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'RefundPaymentIdMissingError';
        this.errorCode = 'REFUND_PAYMENT_ID_MISSING_ERROR';
        this.message = 'REFUND_PAYMENT_ID_MISSING_ERROR';
    }
}
exports.RefundPaymentIdMissingError = RefundPaymentIdMissingError;
class RefundStateTransitionError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'RefundStateTransitionError';
        this.errorCode = 'REFUND_STATE_TRANSITION_ERROR';
        this.message = 'REFUND_STATE_TRANSITION_ERROR';
        this.fromState = input.fromState;
        this.toState = input.toState;
        this.transitionError = input.transitionError;
    }
}
exports.RefundStateTransitionError = RefundStateTransitionError;
class SettlePaymentError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'SettlePaymentError';
        this.errorCode = 'SETTLE_PAYMENT_ERROR';
        this.message = 'SETTLE_PAYMENT_ERROR';
        this.paymentErrorMessage = input.paymentErrorMessage;
    }
}
exports.SettlePaymentError = SettlePaymentError;
const errorTypeNames = new Set(['AlreadyRefundedError', 'CancelActiveOrderError', 'CancelPaymentError', 'ChannelDefaultLanguageError', 'CouponCodeExpiredError', 'CouponCodeInvalidError', 'CouponCodeLimitError', 'CreateFulfillmentError', 'EmailAddressConflictError', 'EmptyOrderLineSelectionError', 'FacetInUseError', 'FulfillmentStateTransitionError', 'GuestCheckoutError', 'IneligibleShippingMethodError', 'InsufficientStockError', 'InsufficientStockOnHandError', 'InvalidCredentialsError', 'InvalidFulfillmentHandlerError', 'ItemsAlreadyFulfilledError', 'LanguageNotAvailableError', 'ManualPaymentStateError', 'MimeTypeError', 'MissingConditionsError', 'MultipleOrderError', 'NativeAuthStrategyError', 'NegativeQuantityError', 'NoActiveOrderError', 'NoChangesSpecifiedError', 'NothingToRefundError', 'OrderLimitError', 'OrderModificationError', 'OrderModificationStateError', 'OrderStateTransitionError', 'PaymentMethodMissingError', 'PaymentOrderMismatchError', 'PaymentStateTransitionError', 'ProductOptionInUseError', 'QuantityTooGreatError', 'RefundOrderStateError', 'RefundPaymentIdMissingError', 'RefundStateTransitionError', 'SettlePaymentError']);
function isGraphQLError(input) {
    return input instanceof ErrorResult || errorTypeNames.has(input.__typename);
}
exports.adminErrorOperationTypeResolvers = {
    AddFulfillmentToOrderResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Fulfillment';
        },
    },
    UpdateOrderItemsResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Order';
        },
    },
    AddManualPaymentToOrderResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Order';
        },
    },
    ApplyCouponCodeResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Order';
        },
    },
    AuthenticationResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'CurrentUser';
        },
    },
    CancelOrderResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Order';
        },
    },
    CancelPaymentResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Payment';
        },
    },
    CreateAssetResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Asset';
        },
    },
    CreateChannelResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Channel';
        },
    },
    CreateCustomerResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Customer';
        },
    },
    CreatePromotionResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Promotion';
        },
    },
    NativeAuthenticationResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'CurrentUser';
        },
    },
    ModifyOrderResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Order';
        },
    },
    RefundOrderResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Refund';
        },
    },
    RemoveOrderItemsResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Order';
        },
    },
    RemoveFacetFromChannelResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Facet';
        },
    },
    RemoveOptionGroupFromProductResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Product';
        },
    },
    SetCustomerForDraftOrderResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Order';
        },
    },
    SetOrderShippingMethodResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Order';
        },
    },
    SettlePaymentResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Payment';
        },
    },
    SettleRefundResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Refund';
        },
    },
    TransitionFulfillmentToStateResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Fulfillment';
        },
    },
    TransitionOrderToStateResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Order';
        },
    },
    TransitionPaymentToStateResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Payment';
        },
    },
    UpdateChannelResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Channel';
        },
    },
    UpdateCustomerResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Customer';
        },
    },
    UpdateGlobalSettingsResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'GlobalSettings';
        },
    },
    UpdatePromotionResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Promotion';
        },
    },
};
