// @vendure/core/dist/common/error/generated-graphql-shop-errors.js
"use strict";
/* eslint-disable */
/** This file was generated by the graphql-errors-plugin, which is part of the "codegen" npm script. */
Object.defineProperty(exports, "__esModule", { value: true });
exports.shopErrorOperationTypeResolvers = exports.VerificationTokenInvalidError = exports.VerificationTokenExpiredError = exports.PaymentFailedError = exports.PaymentDeclinedError = exports.PasswordValidationError = exports.PasswordResetTokenInvalidError = exports.PasswordResetTokenExpiredError = exports.PasswordAlreadySetError = exports.OrderStateTransitionError = exports.OrderPaymentStateError = exports.OrderModificationError = exports.OrderLimitError = exports.NotVerifiedError = exports.NoActiveOrderError = exports.NegativeQuantityError = exports.NativeAuthStrategyError = exports.MissingPasswordError = exports.InvalidCredentialsError = exports.InsufficientStockError = exports.IneligibleShippingMethodError = exports.IneligiblePaymentMethodError = exports.IdentifierChangeTokenInvalidError = exports.IdentifierChangeTokenExpiredError = exports.GuestCheckoutError = exports.EmailAddressConflictError = exports.CouponCodeLimitError = exports.CouponCodeInvalidError = exports.CouponCodeExpiredError = exports.AlreadyLoggedInError = exports.ErrorResult = void 0;
class ErrorResult {
}
exports.ErrorResult = ErrorResult;
class AlreadyLoggedInError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'AlreadyLoggedInError';
        this.errorCode = 'ALREADY_LOGGED_IN_ERROR';
        this.message = 'ALREADY_LOGGED_IN_ERROR';
    }
}
exports.AlreadyLoggedInError = AlreadyLoggedInError;
class CouponCodeExpiredError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'CouponCodeExpiredError';
        this.errorCode = 'COUPON_CODE_EXPIRED_ERROR';
        this.message = 'COUPON_CODE_EXPIRED_ERROR';
        this.couponCode = input.couponCode;
    }
}
exports.CouponCodeExpiredError = CouponCodeExpiredError;
class CouponCodeInvalidError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'CouponCodeInvalidError';
        this.errorCode = 'COUPON_CODE_INVALID_ERROR';
        this.message = 'COUPON_CODE_INVALID_ERROR';
        this.couponCode = input.couponCode;
    }
}
exports.CouponCodeInvalidError = CouponCodeInvalidError;
class CouponCodeLimitError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'CouponCodeLimitError';
        this.errorCode = 'COUPON_CODE_LIMIT_ERROR';
        this.message = 'COUPON_CODE_LIMIT_ERROR';
        this.couponCode = input.couponCode;
        this.limit = input.limit;
    }
}
exports.CouponCodeLimitError = CouponCodeLimitError;
class EmailAddressConflictError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'EmailAddressConflictError';
        this.errorCode = 'EMAIL_ADDRESS_CONFLICT_ERROR';
        this.message = 'EMAIL_ADDRESS_CONFLICT_ERROR';
    }
}
exports.EmailAddressConflictError = EmailAddressConflictError;
class GuestCheckoutError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'GuestCheckoutError';
        this.errorCode = 'GUEST_CHECKOUT_ERROR';
        this.message = 'GUEST_CHECKOUT_ERROR';
        this.errorDetail = input.errorDetail;
    }
}
exports.GuestCheckoutError = GuestCheckoutError;
class IdentifierChangeTokenExpiredError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'IdentifierChangeTokenExpiredError';
        this.errorCode = 'IDENTIFIER_CHANGE_TOKEN_EXPIRED_ERROR';
        this.message = 'IDENTIFIER_CHANGE_TOKEN_EXPIRED_ERROR';
    }
}
exports.IdentifierChangeTokenExpiredError = IdentifierChangeTokenExpiredError;
class IdentifierChangeTokenInvalidError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'IdentifierChangeTokenInvalidError';
        this.errorCode = 'IDENTIFIER_CHANGE_TOKEN_INVALID_ERROR';
        this.message = 'IDENTIFIER_CHANGE_TOKEN_INVALID_ERROR';
    }
}
exports.IdentifierChangeTokenInvalidError = IdentifierChangeTokenInvalidError;
class IneligiblePaymentMethodError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'IneligiblePaymentMethodError';
        this.errorCode = 'INELIGIBLE_PAYMENT_METHOD_ERROR';
        this.message = 'INELIGIBLE_PAYMENT_METHOD_ERROR';
        this.eligibilityCheckerMessage = input.eligibilityCheckerMessage;
    }
}
exports.IneligiblePaymentMethodError = IneligiblePaymentMethodError;
class IneligibleShippingMethodError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'IneligibleShippingMethodError';
        this.errorCode = 'INELIGIBLE_SHIPPING_METHOD_ERROR';
        this.message = 'INELIGIBLE_SHIPPING_METHOD_ERROR';
    }
}
exports.IneligibleShippingMethodError = IneligibleShippingMethodError;
class InsufficientStockError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'InsufficientStockError';
        this.errorCode = 'INSUFFICIENT_STOCK_ERROR';
        this.message = 'INSUFFICIENT_STOCK_ERROR';
        this.order = input.order;
        this.quantityAvailable = input.quantityAvailable;
    }
}
exports.InsufficientStockError = InsufficientStockError;
class InvalidCredentialsError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'InvalidCredentialsError';
        this.errorCode = 'INVALID_CREDENTIALS_ERROR';
        this.message = 'INVALID_CREDENTIALS_ERROR';
        this.authenticationError = input.authenticationError;
    }
}
exports.InvalidCredentialsError = InvalidCredentialsError;
class MissingPasswordError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'MissingPasswordError';
        this.errorCode = 'MISSING_PASSWORD_ERROR';
        this.message = 'MISSING_PASSWORD_ERROR';
    }
}
exports.MissingPasswordError = MissingPasswordError;
class NativeAuthStrategyError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'NativeAuthStrategyError';
        this.errorCode = 'NATIVE_AUTH_STRATEGY_ERROR';
        this.message = 'NATIVE_AUTH_STRATEGY_ERROR';
    }
}
exports.NativeAuthStrategyError = NativeAuthStrategyError;
class NegativeQuantityError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'NegativeQuantityError';
        this.errorCode = 'NEGATIVE_QUANTITY_ERROR';
        this.message = 'NEGATIVE_QUANTITY_ERROR';
    }
}
exports.NegativeQuantityError = NegativeQuantityError;
class NoActiveOrderError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'NoActiveOrderError';
        this.errorCode = 'NO_ACTIVE_ORDER_ERROR';
        this.message = 'NO_ACTIVE_ORDER_ERROR';
    }
}
exports.NoActiveOrderError = NoActiveOrderError;
class NotVerifiedError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'NotVerifiedError';
        this.errorCode = 'NOT_VERIFIED_ERROR';
        this.message = 'NOT_VERIFIED_ERROR';
    }
}
exports.NotVerifiedError = NotVerifiedError;
class OrderLimitError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'OrderLimitError';
        this.errorCode = 'ORDER_LIMIT_ERROR';
        this.message = 'ORDER_LIMIT_ERROR';
        this.maxItems = input.maxItems;
    }
}
exports.OrderLimitError = OrderLimitError;
class OrderModificationError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'OrderModificationError';
        this.errorCode = 'ORDER_MODIFICATION_ERROR';
        this.message = 'ORDER_MODIFICATION_ERROR';
    }
}
exports.OrderModificationError = OrderModificationError;
class OrderPaymentStateError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'OrderPaymentStateError';
        this.errorCode = 'ORDER_PAYMENT_STATE_ERROR';
        this.message = 'ORDER_PAYMENT_STATE_ERROR';
    }
}
exports.OrderPaymentStateError = OrderPaymentStateError;
class OrderStateTransitionError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'OrderStateTransitionError';
        this.errorCode = 'ORDER_STATE_TRANSITION_ERROR';
        this.message = 'ORDER_STATE_TRANSITION_ERROR';
        this.fromState = input.fromState;
        this.toState = input.toState;
        this.transitionError = input.transitionError;
    }
}
exports.OrderStateTransitionError = OrderStateTransitionError;
class PasswordAlreadySetError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'PasswordAlreadySetError';
        this.errorCode = 'PASSWORD_ALREADY_SET_ERROR';
        this.message = 'PASSWORD_ALREADY_SET_ERROR';
    }
}
exports.PasswordAlreadySetError = PasswordAlreadySetError;
class PasswordResetTokenExpiredError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'PasswordResetTokenExpiredError';
        this.errorCode = 'PASSWORD_RESET_TOKEN_EXPIRED_ERROR';
        this.message = 'PASSWORD_RESET_TOKEN_EXPIRED_ERROR';
    }
}
exports.PasswordResetTokenExpiredError = PasswordResetTokenExpiredError;
class PasswordResetTokenInvalidError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'PasswordResetTokenInvalidError';
        this.errorCode = 'PASSWORD_RESET_TOKEN_INVALID_ERROR';
        this.message = 'PASSWORD_RESET_TOKEN_INVALID_ERROR';
    }
}
exports.PasswordResetTokenInvalidError = PasswordResetTokenInvalidError;
class PasswordValidationError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'PasswordValidationError';
        this.errorCode = 'PASSWORD_VALIDATION_ERROR';
        this.message = 'PASSWORD_VALIDATION_ERROR';
        this.validationErrorMessage = input.validationErrorMessage;
    }
}
exports.PasswordValidationError = PasswordValidationError;
class PaymentDeclinedError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'PaymentDeclinedError';
        this.errorCode = 'PAYMENT_DECLINED_ERROR';
        this.message = 'PAYMENT_DECLINED_ERROR';
        this.paymentErrorMessage = input.paymentErrorMessage;
    }
}
exports.PaymentDeclinedError = PaymentDeclinedError;
class PaymentFailedError extends ErrorResult {
    constructor(input) {
        super();
        this.__typename = 'PaymentFailedError';
        this.errorCode = 'PAYMENT_FAILED_ERROR';
        this.message = 'PAYMENT_FAILED_ERROR';
        this.paymentErrorMessage = input.paymentErrorMessage;
    }
}
exports.PaymentFailedError = PaymentFailedError;
class VerificationTokenExpiredError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'VerificationTokenExpiredError';
        this.errorCode = 'VERIFICATION_TOKEN_EXPIRED_ERROR';
        this.message = 'VERIFICATION_TOKEN_EXPIRED_ERROR';
    }
}
exports.VerificationTokenExpiredError = VerificationTokenExpiredError;
class VerificationTokenInvalidError extends ErrorResult {
    constructor() {
        super();
        this.__typename = 'VerificationTokenInvalidError';
        this.errorCode = 'VERIFICATION_TOKEN_INVALID_ERROR';
        this.message = 'VERIFICATION_TOKEN_INVALID_ERROR';
    }
}
exports.VerificationTokenInvalidError = VerificationTokenInvalidError;
const errorTypeNames = new Set(['AlreadyLoggedInError', 'CouponCodeExpiredError', 'CouponCodeInvalidError', 'CouponCodeLimitError', 'EmailAddressConflictError', 'GuestCheckoutError', 'IdentifierChangeTokenExpiredError', 'IdentifierChangeTokenInvalidError', 'IneligiblePaymentMethodError', 'IneligibleShippingMethodError', 'InsufficientStockError', 'InvalidCredentialsError', 'MissingPasswordError', 'NativeAuthStrategyError', 'NegativeQuantityError', 'NoActiveOrderError', 'NotVerifiedError', 'OrderLimitError', 'OrderModificationError', 'OrderPaymentStateError', 'OrderStateTransitionError', 'PasswordAlreadySetError', 'PasswordResetTokenExpiredError', 'PasswordResetTokenInvalidError', 'PasswordValidationError', 'PaymentDeclinedError', 'PaymentFailedError', 'VerificationTokenExpiredError', 'VerificationTokenInvalidError']);
function isGraphQLError(input) {
    return input instanceof ErrorResult || errorTypeNames.has(input.__typename);
}
exports.shopErrorOperationTypeResolvers = {
    UpdateOrderItemsResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Order';
        },
    },
    AddPaymentToOrderResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Order';
        },
    },
    ApplyCouponCodeResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Order';
        },
    },
    AuthenticationResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'CurrentUser';
        },
    },
    NativeAuthenticationResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'CurrentUser';
        },
    },
    RefreshCustomerVerificationResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Success';
        },
    },
    RegisterCustomerAccountResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Success';
        },
    },
    RemoveOrderItemsResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Order';
        },
    },
    RequestPasswordResetResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Success';
        },
    },
    RequestUpdateCustomerEmailAddressResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Success';
        },
    },
    ResetPasswordResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'CurrentUser';
        },
    },
    SetCustomerForOrderResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Order';
        },
    },
    ActiveOrderResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Order';
        },
    },
    SetOrderShippingMethodResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Order';
        },
    },
    TransitionOrderToStateResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Order';
        },
    },
    UpdateCustomerEmailAddressResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Success';
        },
    },
    UpdateCustomerPasswordResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'Success';
        },
    },
    VerifyCustomerAccountResult: {
        __resolveType(value) {
            return isGraphQLError(value) ? value.__typename : 'CurrentUser';
        },
    },
};
